tests/agents/json-specs/span_types.json

```diff
@@ -0,0 +1,200 @@
+    "<type>.allow_null_subtype": "true to allow null subtype, false by default if omitted",
+    "<type>.allow_unlisted_subtype": "true to allow unlisted subtypes, false by default if omitted",
+    "<type>.subtypes": "root element for sub-types of type '<type>', if omitted or empty subtype must be null, unless 'allow_unlisted_subtype' is set to true",
+    "<type>.subtypes.<subtype>": "sub-type element for <subtype>",
+    "<type>.subtypes.<subtype>.__description": "description of <subtype> subtype (optional)",
+    "<type>.subtypes.<subtype>.__used_by": "list of agents that use <subtype> to help document alignment (optional)"
+  },
+  "app": {
+    "allow_null_subtype": true,
+    "subtypes": {
+      "inferred": {
+        "__description": "Sampling profiler inferred spans",
+        "__used_by": [
+          "java"
+        ]
+      }
+    }
+  },
+  "custom": {
+    "__description": "API custom instrumentation",
+    "__used_by": [
+      "java"
+    ],
+    "allow_null_subtype": true
+  },
+  "db": {
+    "__description": "database span",
+    "subtypes": {
+      "cassandra": {
+        "__description": "Cassandra",
+        "__used_by": [
+          "java"
+        ]
+      },
+      "cosmosdb": {
+        "__description": "Azure CosmosDB"
+      },
+      "db2": {
+        "__description": "IBM DB2",
+        "__used_by": [
+          "java"
+        ]
+      },
+      "derby": {
+        "__description": "Apache Derby",
+        "__used_by": [
+          "java"
+        ]
+      },
+      "dynamodb": {
+        "__description": "AWS DynamoDB"
+      },
+      "elasticsearch": {
+        "__description": "Elasticsearch",
+        "__used_by": [
+          "java"
+        ]
+      },
+      "h2": {
+        "__description": "H2",
+        "__used_by": [
+          "java"
+        ]
+      },
+      "hsqldb": {
+        "__description": "HSQLDB",
+        "__used_by": [
+          "java"
+        ]
+      },
+      "ingres": {
+        "__description": "Ingres"
+      },
+      "mariadb": {
+        "__description": "MariaDB",
+        "__used_by": [
+          "java"
+        ]
+      },
+      "mongodb": {
+        "__description": "MongoDB",
+        "__used_by": [
+          "java"
+        ]
+      },
+      "mysql": {
+        "__description": "MySQL",
+        "__used_by": [
+          "java"
+        ]
+      },
+      "oracle": {
+        "__description": "Oracle Database",
+        "__used_by": [
+          "java"
+        ]
+      },
+      "postgresql": {
+        "__description": "PostgreSQL"
+      },
+      "redis": {
+        "__description": "Redis",
+        "__used_by": [
+          "java"
+        ]
+      },
+      "sqlserver": {
+        "__description": "Microsoft SQL Server",
+        "__used_by": [
+          "java"
+        ]
+      },
+      "unknown": {
+        "__description": "Unknown database",
+        "__used_by": [
+          "java"
+        ]
+      }
+    }
+  },
+  "external": {
+    "subtypes": {
```

Comment:
Shouldn't external have `allow_unlisted_subtype: "true"` in order to allow custom values for custom/proprietary protocols?

Reply:
The goal of this specification is to align all agents on span type/sub-type, at least for the common protocols that are being instrumented.

Checks against this specification is enabled by default, but can be opted-out on a per-test basis, at least in the case of the Java agent, thus for custom protocols that do not fit the spec (or for very open-ended cases like using the API), testing should opt-out for those checks when they don't fit.

Reply:
Ok, so this is not limiting any processing/visualization in apm-server/kibana?

Reply:
Not at all, it does not add any constraint on the protocol itself. What we aim for here is to have all agents aligned. So, for example it allows to ensure that two distinct agents on two different platforms like Ruby and Java both use `db` / `mysql` when they are instrumenting calls to a `MySQL` database.

Reply:
Ok, Thanks you for the info